### PR TITLE
Add FollowSymlinks option

### PR DESCRIPTION
**Pull request by calmh**
This adds symlink following. It seems to work, but there's a pretty serious caveat about reconfiguring FollowSymlinks true -> false that basically deletes all your files unless you're careful... (I'll write some docs) I'm thinking about getting this in there without a UI option, as a hidden test feature for a release or two before letting it be widely known...

### CODE REVIEW

**Comment by Zillode** [comment]
@st-jenkins run integration tests for windows please

**Comment by calmh** [comment]
http://docs.syncthing.net/draft/followsymlinks.html

**Comment by calmh** [comment]
@st-jenkins run integration tests please

**Comment by calmh** [comment]
Also versioning may have some difficulty with this. So I guess the questions are "does this make sense", "can we make it good enough" and "do we really want this"...

File: internal/scanner/walk.go [hunk 1]
```diff
@@ -217,9 +231,48 @@ func (w *Walker) walkAndHashFiles(fchan chan protocol.FileInfo) filepath.WalkFun
...
+							l.Debugln(p, "is covered by", dir)
+						}
+						return skip
+					}
+				}
+
+				// The slash after the path name is essential; it makes
+				// filepath.Walk not see the symlink, but the actual
+				// directory on the other side.
+				w.dirsToScan = append(w.dirsToScan, p+"/")
```
Comment (Zillode) [inline internal/scanner/walk.go]: os.PathSeparator

File: internal/scanner/walk.go [hunk 2]
```diff
@@ -97,6 +101,9 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
 	newParallelHasher(w.Dir, w.BlockSize, w.Hashers, hashedFiles, files)
 
 	go func() {
+		w.symlinksSeen = make(map[string]struct{})
```
Comment (Zillode) [inline internal/scanner/walk.go]: unused

File: internal/scanner/walk.go [hunk 3]
```diff
@@ -97,6 +101,9 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
...
+		w.dirsToScan = nil
```
Comment (Zillode) [inline internal/scanner/walk.go]: We are sure that no two Walk() operations can happen at the same time, correct?
Seems like a bad smell to let this information change within a go routine. I'd pass this array as a second return value.

File: test/symlink_test.go [hunk 4]
```diff
@@ -79,11 +83,114 @@ func TestSymlinksStaggeredVersioning(t *testing.T) {
...
+			t.Error(file, "should be file")
+		}
+	}
+
+	for _, nonex := range []string{"s2/d1/d2", "s2/d1/d3"} {
+		_, err := os.Lstat(nonex)
+		if !os.IsNotExist(err) {
+			t.Error(nonex, "should not exist")
+		}
+	}
```
Comment (Zillode) [inline test/symlink_test.go]: Add a test to verify if s1/d1 is still a symlink after syncing
```info, err = os.Stat("s1/d1"); info.Mode()&os.ModeSymlink == os.ModeSymlink```

File: internal/scanner/walk.go (hunk 3 above)
Comment (calmh) [inline internal/scanner/walk.go]: Yeah, it's ugly. We can't affect the return value of filepath.Walk so our options are a closure over the walk func or an attribute like now. Probably the closure is better, as we already do that for the channel and so on. I had a different thing going on at the beginning. :)

File: test/symlink_test.go (hunk 4 above)
Comment (calmh) [inline test/symlink_test.go]: Added

**Comment by calmh** [comment]
What do you all think here? Conceptually it's fairly simple, but the effect of switching the option off (all files previously seen getting delete records) is kinda bad. Can we fix that somehow, and if not, should this go in somehow anyway or not?

**Comment by AudriusButkevicius** [comment]
I think it's fine as it is.
Stuff that would get deleted would be a copy of something else anyway.

**Comment by plouj** [comment]
> What do you all think here? Conceptually it's fairly simple, but the effect of switching the option off (all files previously seen getting delete records) is kinda bad. Can we fix that somehow, and if not, should this go in somehow anyway or not?

Why not make this an option only when creating a folder so that if someone really wants to disable it they will have to create a new folder?

**Comment by calmh** [comment]
Ok, I'll test and polish a little then. I think what happens index wise when the option is turned off, is that

- foo (dir)
- foo/bar (symlink, followed, announced as dir)
- foo/bar/baz (file)

Gets changed to

- foo (dir)
- foo/bar (symlink)
- foo/bar/baz (file)

Baz is untouched, as it's still reachable via the symlink. The other side removes the foo/bar directory and replaces it with the symlink (how...? Must check). The other side no longer sees foo/bar/baz and sends a delete for it. It gets deleted from the source via the symlink.

**Comment by AudriusButkevicius** [comment]
But baz is still available somehwere, which is most important. Yeah not sure there is a way around this, perhaps the flag should be in the index + config so that when someone nukes index we have it in the config, but when someone changes it in the config we error the folder out asking it to be re-added from scratch... But that's just a preventative measure.

**Comment by calmh** [comment]
Actually no, it gets nuked everywhere. I need to get to the bottom of why, I'll be back afterwards. :)

**Comment by calmh** [comment]
There's another fairly bad scenario that I'm not sure I like...

Device A: FollowSymlinks=false
Device B: FollowSymlinks=true

`A` creates a symlink `foo -> /etc`. This gets synced to `B`, who creates the symlink identically. Next scan, `B` follows the symlink. `foo` gets converted into a directory entry with the full contents of `/etc/`. `A` gets a copy of `B`s `/etc`.

Should we ignore incoming index updates for symlinks when FollowSymlinks is true?

**Comment by calmh** [comment]
BTW I figured out why the files get deleted as described above. It only happens when the other side has some kind of versioning enabled, because then we *move* the entire directory structure away when replacing it with a symlink. In the absence of versioning we try a regular remove, which fails with an expected "directory not empty".

**Comment by AudriusButkevicius** [comment]
Yeah this is another one of the situations where a setting on one node can cause having in the cluster.
One guy with FollowSymlinks will nuke symlinks of other nodes isn't it?

**Comment by Zillode** [comment]
@st-jenkins run integration tests for windows please

**Comment by Zillode** [comment]
@st-jenkins run integration tests for windows and osx please

**Comment by calmh** [comment]
I've updated this to merge all the changes v0.11.11 → v0.11.26. That makes it unmergeable of course, but the build should be fine and it's a bit closer to current reality, should we want to continue it.

**Comment by calmh** [comment]
To clarify my position on this, I'm still not convinced that doing this is a good idea. It causes some unintuitive problems that will bite people and as far as I'm concerned the functionality added is niche enough that almost nobody cares. Someone who actually does care needs to drive this.

**Comment by calmh** [comment]
Left to the community to resurrect if someone cares enough to drive it home.

**Comment by Albirew** [comment]
>There's another fairly bad scenario that I don't like...
>
>Device A: FollowSymlinks=false
>Device B: FollowSymlinks=true
>
>A creates a symlink foo -> /etc. This gets synced to B, who creates the symlink identically. Next scan, B follows the symlink. foo gets converted into a directory entry with the full contents of /etc/. A gets a copy of Bs /etc.
>
>Should we ignore incoming index updates for symlinks when FollowSymlinks is true? I think we should...

A simple workaround would be that FollowSymlinks=true announce simlinks as normal directories:

A creates a symlink foo -> /etc. This gets synced to B, who creates an empty foo directory. The end.
As for the reverse:
B creates a symlink foo -> /etc. This gets synced to A, who creates a foo directory with the full contents of B's /etc/. The end.

False FollowSymlinks ≠ CopySymlinks

Another workaround can be to make the FollowSymlinks option global per share so if set to off on a device, it sets to off everywhere else.

**Pull request closed without merging**